# 1.데이터 로드


In [1]:
import pandas as pd
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import folium
from folium import plugins
import json
warnings.filterwarnings(action='ignore')

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [7]:
cd "/content/drive/Shared drives/투미들쓰의 국도교통 해커톤/code/민정미경기차버스"

/content/drive/Shared drives/투미들쓰의 국도교통 해커톤/code/민정미경기차버스


In [6]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [3]:
#폴리움 한글 지원을 위해 설치. 설치이후 런타임 다시시작 // 이 코드는 주석처리
# pip install git+https://github.com/python-visualization/branca.git@master 

## A. 출발지 별 가까운 플랫폼

In [9]:
near = pd.read_csv("구별_가까운곳3.csv")
near

이름                        기차역                                  터미널
0    도봉구  상봉역,청량리역,서울역,용산역,영등포역,수서역  동서울종합터미널,서울고속버스터미널,센트럴시티터미널,서울남부터미널
1    노원구  상봉역,청량리역,서울역,용산역,영등포역,수서역  동서울종합터미널,서울고속버스터미널,센트럴시티터미널,서울남부터미널
2    강북구      청량리역,상봉역,서울역,용산역,영등포역  동서울종합터미널,서울고속버스터미널,센트럴시티터미널,서울남부터미널
3    성북구      청량리역,서울역,상봉역,용산역,영등포역  동서울종합터미널,서울고속버스터미널,센트럴시티터미널,서울남부터미널
4   동대문구      청량리역,상봉역,서울역,용산역,영등포역  동서울종합터미널,서울고속버스터미널,센트럴시티터미널,서울남부터미널
5    중랑구  상봉역,청량리역,서울역,용산역,영등포역,수서역  동서울종합터미널,서울고속버스터미널,센트럴시티터미널,서울남부터미널
6    은평구      서울역,용산역,영등포역,청량리역,상봉역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
7    종로구      서울역,용산역,영등포역,청량리역,상봉역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
8   서대문구      서울역,용산역,영등포역,청량리역,상봉역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
9     중구      서울역,용산역,영등포역,청량리역,상봉역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
10   성동구      서울역,청량리역,용산역,영등포역,상봉역  서울고속버스터미널,센트럴시티터미널,동서울종합터미널,서울남부터미널
11   광진구  수서역,청량리역,상봉역,서울역,용산역,영등포역  동서울종합터미널,서울고속버스터미널,센트럴시티터미널,서울남부터미널
12   용산구  용산역,서울역,영등포역,청량리역,상봉역,수서역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
13   마포구  용산역,영등포역,서울역,청량리역,상봉역,수서역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
14   강동구  수서역,청량리역,상봉역,서울역,용산역,영등포역  동서울종합터미널,서울고속버스터미널,센트럴시티터미널,서울남부터미널
15   송파구  수서역,청량리역,상봉역,서울역,용산역,영등포역  동서울종합터미널,서울고속버스터미널,센트럴시티터미널,서울남부터미널
16   강남구  수서역,서울역,용산역,청량리역,상봉역,영등포역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
17   서초구  용산역,서울역,영등포역,수서역,청량리역,상봉역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
18   동작구  영등포역,용산역,서울역,청량리역,상봉역,수서역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
19   관악구  영등포역,용산역,서울역,청량리역,상봉역,수서역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
20  영등포구  영등포역,용산역,서울역,청량리역,상봉역,수서역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
21   금천구  영등포역,용산역,서울역,청량리역,상봉역,수서역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
22   양천구  영등포역,용산역,서울역,청량리역,상봉역,수서역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
23   구로구  영등포역,용산역,서울역,청량리역,상봉역,수서역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널
24   강서구  영등포역,용산역,서울역,청량리역,상봉역,수서역  서울고속버스터미널,센트럴시티터미널,서울남부터미널,동서울종합터미널

## B. 도착지별 플랫폼 구분

In [10]:
dest = pd.read_csv("도착지터미널구분.csv")[:15]
dest

도착지  ...                     기차
0     부산  ...       서울역,용산역,영등포역,수서역
1     대구  ...       서울역,용산역,영등포역,수서역
2     인천  ...                    NaN
3     광주  ...                서울역,용산역
4     대전  ...           서울역,용산역,영등포역
5     울산  ...       서울역,용산역,영등포역,수서역
6    경기도  ...           서울역,용산역,청량리역
7    강원도  ...           서울역,청량리역,상봉역
8   충청북도  ...  서울역,용산역,영등포역,청량리역,수서역
9   충청남도  ...       서울역,용산역,영등포역,수서역
10  전라북도  ...            서울역,용산역,수서역
11  전라남도  ...            서울역,용산역,수서역
12  경상북도  ...  서울역,용산역,영등포역,청량리역,수서역
13  경상남도  ...       서울역,용산역,영등포역,수서역
14    세종  ...                    NaN

[15 rows x 4 columns]

## C. new_발권OD

In [11]:
bo = pd.read_csv("new_발권OD.csv")

In [12]:
bo

년도      년월     운행 일자  ...  도착 시도 코드 도착 시군구 코드    도착 지역 코드
0        2019  201901  20190101  ...        11     11230  1123010400
1        2019  201901  20190101  ...        11     11260  1126010200
2        2019  201901  20190101  ...        11     11560  1156010100
3        2019  201901  20190101  ...        26     26170  2617010100
4        2019  201901  20190101  ...        26     26320  2632010200
...       ...     ...       ...  ...       ...       ...         ...
3404048  2020  202005  20200522  ...        48     48330     4833010
3404049  2020  202005  20200522  ...        48     48860     4886036
3404050  2020  202005  20200522  ...        48     48860     4886038
3404051  2020  202005  20200522  ...        99     99999  9999999999
3404052  2020  202005  20200522  ...         ~         ~           ~

[3404053 rows x 19 columns]

## D. 서울시_주요_기차역+터미널_좌표


In [13]:
gps_plat = pd.read_csv("서울시_주요_기차역+터미널_좌표.csv")
gps_plat

이름         위도          경도
0        서울역  37.556030  126.972248
1        용산역  37.530186  126.964812
2        수서역  37.487496  127.101324
3       영등포역  37.515609  126.907129
4       청량리역  37.581548  127.048974
5        상봉역  37.596801  127.084762
6  서울고속버스터미널  37.505612  127.007071
7   동서울종합터미널  37.533601  127.093367
8    서울남부터미널  37.484465  127.016054
9   센트럴시티터미널  37.505080  127.003227

In [14]:
#구별로 영역을 나누기 위한 파일
geo_path = '02. skorea_municipalities_geo_simple.json'
geo_str = json.load(open(geo_path, encoding='utf-8'))

#기본 맵 정의
map = folium.Map(location=[37.5502, 126.982], zoom_start=12,
                tiles='stamentoner')
map.choropleth(geo_data=geo_str, fill_color = 'gray')

train = gps_plat.loc[:5]
minal = gps_plat.loc[6:]
minal.reset_index(drop=True, inplace=True)

#기차역 좌표 추가
for i in range(len(train)):
  x = train['위도'][i]
  y = train['경도'][i]
  folium.Marker([x, y], tooltip=train['이름'][i], popup=train['이름'][i]).add_to(map)

#버스터미널 좌표 추가
for i in range(len(minal)):
  x = minal['위도'][i]
  y = minal['경도'][i]
  folium.Marker([x, y], popup=minal['이름'][i], tooltip=minal['이름'][i], icon=folium.Icon(color='green', icon='star')).add_to(map)

map

## E. 서울시 행정구역 시군구 정보

In [15]:
gps_gu = pd.read_csv("서울시 행정구역 시군구 정보 (좌표계_ WGS1984).csv", encoding='cp949')
gps_gu

순번  시군구코드 시군구명_한글          시군구명_영문  ESRI_PK         위도          경도
0    1  11320     도봉구        Dobong-gu        0  37.665861  127.031767
1    2  11380     은평구     Eunpyeong-gu        1  37.617612  126.922700
2    3  11230    동대문구    Dongdaemun-gu        2  37.583801  127.050700
3    4  11590     동작구       Dongjak-gu        3  37.496504  126.944307
4    5  11545     금천구     Geumcheon-gu        4  37.460097  126.900155
5    6  11530     구로구          Guro-gu        5  37.495486  126.858121
6    7  11110     종로구        Jongno-gu        6  37.599100  126.986149
7    8  11305     강북구       Gangbuk-gu        7  37.646995  127.014716
8    9  11260     중랑구      Jungnang-gu        8  37.595379  127.093967
9   10  11680     강남구       Gangnam-gu        9  37.495985  127.066409
10  11  11500     강서구       Gangseo-gu       10  37.565762  126.822656
11  12  11140      중구          Jung-gu       11  37.557945  126.994190
12  13  11740     강동구      Gangdong-gu       12  37.549208  127.146482
13  14  11215     광진구      Gwangjin-gu       13  37.548144  127.085753
14  15  11440     마포구          Mapo-gu       14  37.562291  126.908780
15  16  11650     서초구        Seocho-gu       16  37.476953  127.037810
16  17  11290     성북구      Seongbuk-gu       17  37.606991  127.023218
17  18  11350     노원구         Nowon-gu       18  37.655264  127.077120
18  19  11710     송파구        Songpa-gu       19  37.504853  127.114482
19  20  11410    서대문구     Seodaemun-gu       21  37.582037  126.935667
20  21  11470     양천구     Yangcheon-gu       22  37.527062  126.856153
21  22  11560    영등포구  Yeongdeungpo-gu       23  37.520641  126.913924
22  23  11620     관악구        Gwanak-gu       15  37.465399  126.943807
23  24  11200     성동구     Seongdong-gu       20  37.550675  127.040962
24  25  11170     용산구       Yongsan-gu       24  37.531101  126.981074

# 2. 발권OD 데이터 전처리 (서울특별시만, 기간수정)

In [16]:
bo_seoul = bo[bo['출발 시도 명'] == "서울특별시"]
bo_seoul.dropna(subset=['도착 시도 명', '도착 시군구 명'], inplace=True)
bo_seoul

년도      년월     운행 일자  ...  도착 시도 코드 도착 시군구 코드    도착 지역 코드
0        2019  201901  20190101  ...        11     11230  1123010400
1        2019  201901  20190101  ...        11     11260  1126010200
2        2019  201901  20190101  ...        11     11560  1156010100
3        2019  201901  20190101  ...        26     26170  2617010100
4        2019  201901  20190101  ...        26     26320  2632010200
...       ...     ...       ...  ...       ...       ...         ...
3394200  2020  202005  20200522  ...        42     42110     4211034
3394201  2020  202005  20200522  ...        42     42110     4211040
3394202  2020  202005  20200522  ...        44     44131     4413111
3394203  2020  202005  20200522  ...        44     44200     4420011
3394204  2020  202005  20200522  ...        44     44200     4420033

[224693 rows x 19 columns]

In [17]:
bo_seoul_2019 = bo_seoul[bo_seoul['년도'] == 2019]
bo_seoul_2020 = bo_seoul[bo_seoul['년도'] == 2020]
bo_seoul_2020

년도      년월     운행 일자  ...  도착 시도 코드 도착 시군구 코드    도착 지역 코드
1808899  2020  202001  20200101  ...        45     45180  4518010500
1808907  2020  202001  20200101  ...        42     42210     4221010
1808941  2020  202001  20200101  ...        41     41820  4182025023
1808945  2020  202001  20200101  ...        42     42130  4213032000
1808978  2020  202001  20200101  ...        41     41171  4117110100
...       ...     ...       ...  ...       ...       ...         ...
3394200  2020  202005  20200522  ...        42     42110     4211034
3394201  2020  202005  20200522  ...        42     42110     4211040
3394202  2020  202005  20200522  ...        44     44131     4413111
3394203  2020  202005  20200522  ...        44     44200     4420011
3394204  2020  202005  20200522  ...        44     44200     4420033

[112732 rows x 19 columns]

In [18]:
#첫 확진자 이후인 1월 19일 데이터를 기준으로 보기 위하여 한번 더 분리 (분석에 이용할 최종 데이터이므로 od_19/od_20으로 명명)
od_19 = bo_seoul_2019[bo_seoul_2019['운행 일자'] > 20190118]
od_20 = bo_seoul_2020[bo_seoul_2020['운행 일자'] > 20200118]

In [19]:
od_19.reset_index(inplace=True, drop=True)
od_19

년도      년월     운행 일자  요일 구분  ...    출발 지역 코드 도착 시도 코드 도착 시군구 코드    도착 지역 코드
0      2019  201901  20190120      1  ...  1117010700       11     11170  1117012800
1      2019  201901  20190120      1  ...  1117010700       11     11230  1123010400
2      2019  201901  20190120      1  ...  1117010700       11     11260  1126010200
3      2019  201901  20190120      1  ...  1117010700       11     11560  1156010100
4      2019  201901  20190120      1  ...  1117010700       26     26170  2617010100
...     ...     ...       ...    ...  ...         ...      ...       ...         ...
98663  2019  201905  20190523      5  ...  1168011500       47     47150  4715032021
98664  2019  201905  20190523      5  ...     1171010       42     42110     4211012
98665  2019  201905  20190523      5  ...     1171010       42     42110     4211034
98666  2019  201905  20190523      5  ...     1171010       42     42110     4211040
98667  2019  201905  20190523      5  ...     1171010       44     44131     4413111

[98668 rows x 19 columns]

In [20]:
od_20.reset_index(drop=True, inplace=True)
od_20

년도      년월     운행 일자  요일 구분  ...    출발 지역 코드 도착 시도 코드 도착 시군구 코드    도착 지역 코드
0      2020  202001  20200119      1  ...  1117010700       11     11170  1117012800
1      2020  202001  20200119      1  ...  1117010700       11     11230  1123010400
2      2020  202001  20200119      1  ...  1117010700       11     11260  1126010200
3      2020  202001  20200119      1  ...  1117010700       11     11560  1156010100
4      2020  202001  20200119      1  ...  1117010700       26     26170  2617010100
...     ...     ...       ...    ...  ...         ...      ...       ...         ...
98307  2020  202005  20200522      6  ...     1171010       42     42110     4211034
98308  2020  202005  20200522      6  ...     1171010       42     42110     4211040
98309  2020  202005  20200522      6  ...     1171010       44     44131     4413111
98310  2020  202005  20200522      6  ...     1171010       44     44200     4420011
98311  2020  202005  20200522      6  ...     1171010       44     44200     4420033

[98312 rows x 19 columns]

# 3. 기차역, 터미널 코드

<기차역>  
[ 출발 지역 코드 - 지역 - 기차역 ]  
'1117010700' - 용산구 동자동 - 서울역  
'1117012800' - 용산구 한강로3가 - 용산역  
'1120010700' - 성동구 행당동 - 왕십리역  
'1120011300' - 성동구 옥수동 - 옥수역  
'1123010400' - 동대문구 전농동 - 청량리역  
'1126010200' - 중랑구 상봉동 - 상봉역  
'1156010100' - 영등포구 영등포동 - 영등포역  
'1168011500' - 강남구 수서동 - 수서역  

[코드 -> 행정동 조회](https://www.code.go.kr/stdcode/regCodeL.do)  
[행정동 -> 코드 조회](http://minwon.gyeongnam.go.kr/citynet/jsp/cmm/CMMAddrRegnCodeAction.do)  

<버스 터미널>  
[ 출발 지역 코드 - 지역 - 버스 터미널 ]  
1165010700 - 서초구 반포동 - 서울고속버스터미널 / 센트럴시티터미널 ...①  
1121510 - 광진구  
1165010 - 서초구  
1150010 - 강서구  
1121510300 - 광진구 구의동 - 동서울종합터미널 ...②  
1165010800 - 서초구 서초동 - 서울남부터미널 ...③  
1171010100 - 송파구 잠실동 - 잠실역 시외버스 정류소 ...④  
1171010900 - 송파구 장지동 - 장지역 시외버스 정류소 ...⑤  
1171010700 - 송파구 가락동 - 가락시장 시외버스 정류소 ...⑥  
1171010 - 송파구  
1126010200 - 중랑구 상봉동 - 상봉터미널...⑦  

# --- 최종 함수 정의 ---

In [21]:
def to_query(location, loc):
  start = "출발지역코드 == "
  col = "'"
  loc_code = location[loc]
  return (start + col + loc_code + col), loc_code

In [22]:
def find_res_plat(dest, near):
  res = ""
  for i in range(len(near)):
    if near[i] in dest:
      res += near[i]
      break
  return res

In [62]:
from IPython.display import display
# LDN_COORDINATES = (51.5074, 0.1278) 

def view_map(start_region, res_plat):

  #한글 셋팅 및 시군구 분리
  plt.rc('font', family='NanumBarunGothic') 
  geo_path = '02. skorea_municipalities_geo_simple.json'
  geo_str = json.load(open(geo_path, encoding='utf-8'))

  #기본 맵 정의
  f = folium.Figure(width=700, height=400)
  map = folium.Map(location=[37.5502, 126.982], zoom_start=11, tiles='stamentoner').add_to(f)
  map.choropleth(geo_data=geo_str, fill_color = 'skyblue')

  #출발지 빨간색으로 체크
  gu_xy = gps_gu[gps_gu['시군구명_한글'] == start_region][['위도', '경도']].values
  gu_x = gu_xy[0][0]
  gu_y = gu_xy[0][1]
  folium.Marker([gu_x, gu_y], popup=start_region, tooltip=start_region, icon=folium.Icon(color='red', icon='star')).add_to(map)

  #최종 선정된 터미널역 
  for i in range(len(res_plat)):
    plat_xy = gps_plat[gps_plat['이름'] == res_plat[i]][['위도', '경도']].values
    plat_x = plat_xy[0][0]
    plat_y = plat_xy[0][1]
    folium.Marker([plat_x, plat_y], tooltip=res_plat[i], popup=res_plat[i]).add_to(map)

  display(map)
  return map

In [63]:
def find_platform(df, start_region, destination):

  #필요한 정보들
  df.rename(columns={"출발 지역 코드":"출발지역코드"}, inplace=True)
  near = pd.read_csv("구별_가까운곳3.csv")
  dest = pd.read_csv("도착지터미널구분.csv")[:15]
  location = {'서울역' : '1117010700', '용산역' : '1117012800', '수서역' : '1168011500', '청량리역' : '1123010400', '영등포역' : '1156010100', '상봉역' : '1126010200',
            '서울고속버스터미널' : '1165010700', '서울남부터미널' : '1165010800', '동서울종합터미널' : '1121510300', '센트럴시티터미널' : '1165010700'}
  

  print("<< 출발지", start_region, "- 도착지", destination, "으로 검색합니다. >> \n")

  #1. 도착지로 운행하는 플랫폼 목록 (express: 고속버스, inter: 시외버스, train: 기차역)
  dest_list = dest[dest['도착지'] == destination]
  try:
    dest_express = dest_list['고속버스'].values[0].split(",")
  except:
    dest_express = []
  try:
    dest_inter = dest_list['시외버스'].values[0].split(",")
  except:
    dest_inter = []
  try:  
    dest_train = dest_list['기차'].values[0].split(",")
  except:
    dest_train = []

  # dest_list = dest[dest['도착지'] == destination]
  # dest_express = dest_list['고속버스'].values[0].split(",")
  # dest_inter = dest_list['시외버스'].values[0].split(",")
  # dest_train = dest_list['기차'].values[0].split(",")

  #도착지로 운행하는 교통수단에 대한 안내 메시지
  dest_ava_list = []
  if len(dest_express) >= 1:
    dest_ava_list.append("고속버스")
  if len(dest_inter) >= 1:
    dest_ava_list.append("시외버스")
  if len(dest_train) >= 1:
    dest_ava_list.append("기차")
  print(destination,"까지 운행 가능한 교통수단은", dest_ava_list, "입니다.\n")

  #2. 주요 터미널/기차역 중 출발지와 가까운 순서로 정렬한 것.
  near_list = near[near['이름'] == start_region]
  near_terminal = near_list['터미널'].values[0].split(",")
  near_train = near_list['기차역'].values[0].split(",")
  print("출발지에서 가까운 플랫폼은 다음과 같습니다.\n", "기차역 : ", near_train, "\n 터미널 : ", near_terminal, "\n")

  #3. 위의 두 데이터를 고려한, 도착지까지 운행하면서 출발지와 가장 가까운 터미널역 선정
  res_plat = []
  res_plat.append(find_res_plat(dest_train, near_train))
  res_plat.append(find_res_plat(dest_express, near_terminal))
  res_plat.append(find_res_plat(dest_inter, near_terminal))
  print("도착지까지 운행하면서, 출발지와 가장 가까운 플랫폼은 다음과 같습니다. (순서대로 기차역, 고속버스, 시외버스)\n", res_plat, "\n")

  #4. 결과로 선정된 플랫폼을 출발로 하는 데이터 행 필터링
  query = ""
  count = 0
  loc_code = []
  for i in range(len(res_plat)):
    if res_plat[i] != '':
      q, code = to_query(location, res_plat[i])
      query += q
      count += 1
      loc_code.append(code)
      # loc_code.append(code)
      if (i != len(res_plat)-1):
        query += " or "

  #예외처리규정
  if query[-3:] == 'or ':
    query = query[:-4]

  print("최종 쿼리문 \n", "<< ", query, " >> \n")
  map = view_map(start_region, res_plat)

  
  return df.query(query), map, loc_code

In [64]:
df, map, loc_codes = find_platform(od_20, "마포구", "울산")

<< 출발지 마포구 - 도착지 울산 으로 검색합니다. >> 

울산 까지 운행 가능한 교통수단은 ['고속버스', '시외버스', '기차'] 입니다.

출발지에서 가까운 플랫폼은 다음과 같습니다.
 기차역 :  ['용산역', '영등포역', '서울역', '청량리역', '상봉역', '수서역'] 
 터미널 :  ['서울고속버스터미널', '센트럴시티터미널', '서울남부터미널', '동서울종합터미널'] 

도착지까지 운행하면서, 출발지와 가장 가까운 플랫폼은 다음과 같습니다. (순서대로 기차역, 고속버스, 시외버스)
 ['용산역', '서울고속버스터미널', '동서울종합터미널'] 

최종 쿼리문 
 <<  출발지역코드 == '1117012800' or 출발지역코드 == '1165010700' or 출발지역코드 == '1121510300'  >> 



In [29]:
map

In [ ]:
df

In [ ]:
loc_codes